File: cpp/build-company/ConstructionObject.cpp
#include "ConstructionObject.hpp"
#include <iostream>

// Реализация конструктора
ConstructionObject::ConstructionObject(const std::string& address, ObjectType objectType, ProjectStatus status)
    : address(address), objectType(objectType), status(status) {
}

// Реализация деструктора
ConstructionObject::~ConstructionObject() {
    std::cout << "Вызван деструктор ConstructionObject для: " << address << std::endl;
}

// Реализация геттеров
std::string ConstructionObject::getAddress() const {
    return address;
}

ObjectType ConstructionObject::getObjectType() const {
    return objectType;
}

ProjectStatus ConstructionObject::getStatus() const {
    return status;
}

// Реализация сеттеров
void ConstructionObject::setAddress(const std::string& address) {
    this->address = address;
}

void ConstructionObject::setObjectType(ObjectType objectType) {
    this->objectType = objectType;
}

void ConstructionObject::setStatus(ProjectStatus status) {
    this->status = status;
    std::cout << "Статус проекта по адресу " << address << " изменен на: "
              << projectStatusToString(status) << std::endl;
}

// Реализация функции отображения информации
void ConstructionObject::displayInfo() const {
    std::cout << "=== Строительный объект ===" << std::endl;
    std::cout << "Адрес: " << address << std::endl;
    std::cout << "Тип: " << objectTypeToString(objectType) << std::endl;
    std::cout << "Статус: " << projectStatusToString(status) << std::endl;
}

File: csharp/build-company/ConstructionObject.cs
namespace BuildCompany;

/// <summary>
/// Класс, представляющий строительный проект/объект
/// </summary>
public class ConstructionObject
{
    private string address;
    private ObjectType objectType;
    private ProjectStatus status;

    /// <summary>
    /// Конструктор
    /// </summary>
    /// <param name="address">Адрес объекта</param>
    /// <param name="objectType">Тип объекта</param>
    /// <param name="status">Статус проекта</param>
    public ConstructionObject(string address, ObjectType objectType, ProjectStatus status)
    {
        this.address = address;
        this.objectType = objectType;
        this.status = status;
    }

    /// <summary>
    /// Получить адрес
    /// </summary>
    public string GetAddress()
    {
        return address;
    }

    /// <summary>
    /// Получить тип объекта
    /// </summary>
    public ObjectType GetObjectType()
    {
        return objectType;
    }

    /// <summary>
    /// Получить статус
    /// </summary>
    public ProjectStatus GetStatus()
    {
        return status;
    }

    /// <summary>
    /// Установить адрес
    /// </summary>
    public void SetAddress(string address)
    {
        this.address = address;
    }

    /// <summary>
    /// Установить тип объекта
    /// </summary>
    public void SetObjectType(ObjectType objectType)
    {
        this.objectType = objectType;
    }

    /// <summary>
    /// Установить статус
    /// </summary>
    public void SetStatus(ProjectStatus status)
    {
        this.status = status;
        Console.WriteLine($"Project status updated to: {status.ToDisplayString()}");
    }

    /// <summary>
    /// Отображение информации об объекте
    /// </summary>
    public void DisplayInfo()
    {
        Console.WriteLine($"[Construction Object]");
        Console.WriteLine($"  Address: {address}");
        Console.WriteLine($"  Type: {objectType.ToDisplayString()}");
        Console.WriteLine($"  Status: {status.ToDisplayString()}");
    }
}
